python-package/lightgbm/callback.py

```diff
@@ -2,14 +2,27 @@
 """Callbacks library."""
 import collections
 from operator import gt, lt
+from typing import Any, Callable, List, Union
 
 from .basic import _ConfigAliases, _log_info, _log_warning
 
 
+class CallbackWithAttributes():
+    """Allows the creation of function attributes without making mypy raise errors."""
+
+    order: int
+    before_iteration: bool
+
+
+def callback_attr_decorator(func: Any) -> CallbackWithAttributes:
+    """Prevents function attributes from raising mypy errors."""
```

Comment:
```suggestion
    """Adds more specific type annotation for callback functions.
    
    This decorator informs type checkers like mypy that callback functions have attributes
    like ``before_iteration`` and ``order``.
    """
```

Could you make this more specific? I think this proposed description makes it clearer what is going on in this code.

Reply:
I see. Thanks for the suggestion; the proposed description makes it much clearer.

Reply:
I saw this linting error from a failed test. 
```./python-package/lightgbm/callback.py:18 in private function `_callback_attr_decorator`:
        D401: First line should be in imperative mood (perhaps 'Add', not 'Adds')```

It seems to me that we should start the docstring with "Adds" like you suggested and not "Add" for a docstring which describes what the function does. I'll change the first word to "Add" like the error message suggested, but I thought I'd make sure that it's really the right thing to do.